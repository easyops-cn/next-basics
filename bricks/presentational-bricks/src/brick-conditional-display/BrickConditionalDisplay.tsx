import React from "react";
import { merge } from "lodash";

import { DataType, isConditionSatisfied } from "@next-libs/cmdb-utils";
import { RuleProps } from "./index";
import cssStyle from "./style.module.css";

export type DisplayType = "label" | "default";

export interface BrickConditionalDisplayProps {
  data: DataType;
  value: number | string;
  rules: RuleProps[];
  type: DisplayType;
}

export function BrickConditionalDisplay(
  props: BrickConditionalDisplayProps
): React.ReactElement {
  const renderDisplay = (
    data: DataType,
    value: number | string,
    rules: RuleProps[]
  ) => {
    if (data === null) {
      return <span className={cssStyle.error}>参数“dataSource”为空</span>;
    } else {
      let display = <div className={cssStyle[props.type]}>{value}</div>;
      rules.forEach((rule) => {
        if (isConditionSatisfied(data, rule.condition)) {
          const style = merge(
            {
              backgroundColor: "rgba(47, 194, 91, 1)",
              color: "rgba(255, 255, 255, 1)",
              borderColor: "rgba(47, 194, 91, 1)",
            },
            rule.style
          );

          display = (
            <div
              className={cssStyle[props.type]}
              style={{
                borderColor: style.borderColor,
                background: style.backgroundColor,
                color: style.color,
              }}
            >
              {rule.label || value}
            </div>
          );
        }
      });
      return display;
    }
  };

  return renderDisplay(props.data, props.value, props.rules);
}
