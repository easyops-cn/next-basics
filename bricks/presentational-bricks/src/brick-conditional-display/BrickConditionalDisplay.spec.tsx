import React from "react";
import { render } from "@testing-library/react";

import { BrickConditionalDisplay } from "./BrickConditionalDisplay";

describe("BrickConditionalDisplay", () => {
  it("should work", () => {
    const value = "130";
    const rules = [
      {
        condition: {
          $lt: 60
        },
        style: {
          backgroundColor: "rgba(252, 80, 67, 1)",
          color: "rgba(255, 255, 255, 1)"
        },
        label: "不及格"
      },
      {
        condition: {
          $gte: 60,
          $lt: 80
        },
        style: {
          backgroundColor: "rgba(255, 162, 53, 1)",
          color: "rgba(255, 255, 255, 1)"
        },
        label: "及格"
      },
      {
        condition: {
          $gte: 80,
          $lt: 90
        },
        style: {
          backgroundColor: "rgba(47, 194, 91, 1)",
          color: "rgba(255, 255, 255, 1)"
        },
        label: "良好"
      },
      {
        condition: {
          $gte: 90
        },
        style: {
          backgroundColor: "rgba(22, 123, 224, 1)",
          color: "rgba(255, 255, 255, 1)"
        },
        label: "优秀"
      }
    ];

    const result = render(
      <BrickConditionalDisplay
        data={100}
        value={value}
        rules={rules}
        type="label"
      />
    );
    const asFragment = result.asFragment;
    expect(asFragment()).toMatchSnapshot();
  });
});
