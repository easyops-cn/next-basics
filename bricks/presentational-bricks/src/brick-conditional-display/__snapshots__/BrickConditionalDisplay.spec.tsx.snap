// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrickConditionalDisplay should work 1`] = `
<DocumentFragment>
  <div
    class="label"
    style="border-color: rgba(47, 194, 91, 1); background: rgb(22, 123, 224); color: rgb(255, 255, 255);"
  >
    优秀
  </div>
</DocumentFragment>
`;
